synapse/storage/databases/state/store.py

```diff
@@ -404,6 +406,124 @@ def _insert_into_cache(
+        self,
+        events_and_context: List[Tuple[EventBase, UnpersistedEventContextBase]],
+        room_id: str,
+        prev_group: int,
+    ) -> List[Tuple[EventBase, UnpersistedEventContext]]:
+        """Generate and store state deltas for a group of events and contexts created to be
+        batch persisted. Note that all the events must be in a linear chain (ie a <- b <- c)
+        and must be state events.
+
+        Args:
+            events_and_context: the events to generate and store a state groups for
+            and their associated contexts
+            room_id: the id of the room the events were created for
+            prev_group: the state group of the last event persisted before the batched events
+            were created
+        """
+
+        def insert_deltas_group_txn(
+            txn: LoggingTransaction,
+            events_and_context: List[Tuple[EventBase, UnpersistedEventContext]],
+            prev_group: int,
+        ) -> List[Tuple[EventBase, UnpersistedEventContext]]:
+            """Generate and store state groups for the provided events and contexts.
+
+            Requires that we have the state as a delta from the last persisted state group.
+
+            Returns:
+                A list of state groups
+            """
+            is_in_db = self.db_pool.simple_select_one_onecol_txn(
+                txn,
+                table="state_groups",
+                keyvalues={"id": prev_group},
+                retcol="id",
+                allow_none=True,
+            )
+            if not is_in_db:
+                raise Exception(
+                    "Trying to persist state with unpersisted prev_group: %r"
+                    % (prev_group,)
+                )
+
+            num_state_groups = len(events_and_context)
+
+            state_groups = self._state_group_seq_gen.get_next_mult_txn(
+                txn, num_state_groups
+            )
+
+            for index, (event, context) in enumerate(events_and_context):
+                context.state_group_after_event = state_groups[index]
+                # The first prev_group will be the last persisted state group, which is passed in
+                # else it will be the group most recently assigned
+                if index > 0:
+                    context.prev_group_for_state_group_after_event = state_groups[
+                        index - 1
+                    ]
+                    context.state_group_before_event = state_groups[index - 1]
+                else:
+                    context.prev_group_for_state_group_after_event = prev_group
+                    context.state_group_before_event = prev_group
+                context.delta_ids_to_state_group_after_event = {
+                    (event.type, event.state_key): event.event_id
+                }
+                context.state_delta_due_to_event = {
+                    (event.type, event.state_key): event.event_id
+                }
```

Comment:
I think we might be able to simplify this logic, once its been updated to match the changes in the base PR.

Basically what we are aiming to produce is a chain of state groups: `s1 -> s2 -> ... -> sN` and then fill out each contexts' prev group and state groups with adjacent state groups from the sequence, i.e. the first context would likely have `prev_group=None, sg_before=s1, sg_after=s2`, then the second would be `prev_group=s1, sg_before=s2, sg_after=s3`, etc.

So instead of tracking indices, you could have a set of variables `sg_prev, sg_before, sg_after` which are updated at every iteration (`sg_prev` becomes `sg_before`, etc). 

I *think* that makes it easier to follow what's going on?